Assets/Scripts/GameManager.cs

```diff
@@ -0,0 +1,155 @@
+        [SerializeField] [Range(34, 66)] private int probmedium;
+        [SerializeField] [Range(67, 100)] private int problarge;
+        private Options options;
+        private bool gameActive;
+        private UnityEvent onRestartClick;
+        private UnityEvent onGameActiveToggle;
+        private UnityEvent onProbChanged;
+        private UnityEvent onSizeChanged;
+        private UnityEvent onSpeedChanged;
+        private int generation = 1;
+        private int aliveCounter;
+
+        void Start()
+        {
+            options = ScriptableObject.CreateInstance<Options>();
+            options.InitInputsQueues(speedlarge, speedmedium, speedsmall, sizelarge, sizemedium, sizesmall, problarge,
+                probmedium, probsmall);
+            InitEventListeners();
+            restartButton.Render("Restart", onRestartClick,gameActive);
+            renderToggle(ValuesEnum.Prob);
+            renderToggle(ValuesEnum.Size);
+            renderToggle(ValuesEnum.Speed);
+            grid.InitGrid(getMaximalSize(), getProb());
+        }
+
+        void InitEventListeners()
+        {
+            onSizeChanged = new UnityEvent();
+            onSpeedChanged = new UnityEvent();
+            onProbChanged = new UnityEvent();
+            onGameActiveToggle = new UnityEvent();
+            onRestartClick = new UnityEvent();
+            onSizeChanged.AddListener(() => UpdateParam(ValuesEnum.Size));
+            onSpeedChanged.AddListener(() => UpdateParam(ValuesEnum.Speed));
+            onProbChanged.AddListener(() => UpdateParam(ValuesEnum.Prob));
+            onGameActiveToggle.AddListener(() =>
+            {
+                gameActive = !gameActive;
+                if (gameActive) StartCoroutine(GameFunction());
+                renderToggle(ValuesEnum.Prob);
+                renderToggle(ValuesEnum.Size);
+            });
+            onRestartClick.AddListener(() =>
+            {
+                generation = 1;
+                grid.OnRestart(getSize(),getProb());
+            });
+        }
+
+         int getMaximalSize()
+        {
+            return sizelarge;
+        }
+
+        void UpdateParam(ValuesEnum valueToChange)
+        {
+            if (valueToChange == ValuesEnum.Size)
+            {
+                int oldSize = options.GetCurrentValue(ValuesEnum.Size);
+                options.GetNextValue(ValuesEnum.Size);
+                grid.UpdateGridBySize(getSize(),oldSize);
+            }
+            else
+            {
+                options.GetNextValue(valueToChange);
+            }
+
+            if (valueToChange != ValuesEnum.Speed)
+            {
+                generation = 1;
+                grid.OnRestart(getSize(),getProb());
+            }
+
+            renderToggle(valueToChange);
+        }
+
+
+        IEnumerator GameFunction()
+        {
+            while (gameActive)
+            {
+                yield return new WaitForSeconds(getSpeed() / 1000);
```

Comment:
it will be better to cache the WaitForSeconds if the speed didn't change you can use the same object without doing new WaitForSeconds for the same time